### ISSUE TITLE
feat(common): Stable Signature Generation from JS Objects

### ISSUE THREAD

**Issue by petermetz** [hyperledger/cactus#136]
### Description

> I'm not an expert on this topic, might not even be using the right terminology. If the description doesn't make sense I recommend checking out the code example below as that is known to be illustrating the problem in an exact manner whereas my English prose may not.

As a developer I want never ever ever have to struggle with improper canonicalization of data so that I never have to waste my time debugging such a classic/beginner bug.

Also, we absolutely must nip this in the bud early on to make sure that it's not something that we fix "mid-flight" after the first major release that people then deploy to production and start struggling with this (and with the migration later).
This is especially important given the nature of the software that Cactus is (it passes data around to a relatively large number of other software given that it's main purpose is to create bridges between different software).

A good starting point is protocol buffers which represent everything as an array of values (may be an oversimplification, apologies if reader is a PB expert, I'm not).
On the other hand, a problem with protocol buffers on the other hand is the quite heavy handed tooling that (AFAIK) is not always a good match for NodeJS based projects (native addons usually cause problems with build, OS cross-compatibility suffers, webpack code bundling needs workarounds, etc.) 
It's also harder to debug protocol buffers because of the binary encoding whereas JSON designed to be as human readable as possible.
What we need is to find a good trade-off between JSON and protocol buffers (not saying protobuf is out, just that preferably we'd do something that's a better match for the acceptance criteria overall). 

JSON Web Tokens might be a great fit since it seems to solve the same problem even if the goals are slightly different (or maybe they aren't?)
https://jwt.io/introduction/

Or a stable JSON stringify implementation: https://www.npmjs.com/package/json-stable-stringify

### Acceptance Criteria
1. Find a serialization protocol+implementation (library) that makes it impossible to have bugs like property (re)ordering differences producing different signatures for what is essentially the same JS object/JSON string. Example being: 
```javascript
const privateKey = "wouldn't you like to know";

const dataA = JSON.stringify({ walletAddress: "some-wallet-address", ledgerName: "my-cool-ledger" });
const signatureA = cryptoUtils.sign(dataA, privateKey);

const dataB = JSON.stringify({ ledgerName: "my-cool-ledger", walletAddress: "some-wallet-address" });
const signatureB = cryptoUtils.sign(dataB, privateKey);

// This fails because dataA and B has a different order of bytes 
// even though for all of *our* intents and purposes the two are equal...
// So we need something to coerce the two into the same representation
// just before the signature generation.
assert.equals(signatureA, signatureB);
``` 
2. Provide a list of alternatives considered and include a matrix of features/pros/cons to support the decision made about which protocol/implementation to use for future reference (this is a fundamental piece of the codebase because generating signatures for serialized data is a fundamental piece of what we are trying to achieve)
3. Cross-platform implementation e.g. it works the same way in the browser and in NodeJS (this is a must have)
4. Have a class in the `cactus-common` package that is able to pass the tests outlined here and works in both browsers  and NodeJS (making runtime decisions based on platforms to have fallback implementations is acceptable, but it has to be done transparently to the caller side).
5. The class mentioned above has to have a 100% test coverage  (it's a really important part of the code, we are not usually this strict about this, not even close) 
6. As close to good ol' JSON as possible, but without the issue that's being presented here
7. No native NodeJS addons strongly preferred
8. It should be easy to debug the contents. 
For example nested serialization makes it harder to read things as a human in the case of JSON where you do something like this: 
```javascript
const inner = { someProperty: "cool", otherStuff: "also cool"};
const outer = { innerProperty: JSON.stringify(inner) };
JSON.stringify(outer, null, 4);
> "{
    "innerProperty": "{\"someProperty\":\"cool\",\"otherStuff\":\"also cool\"}"
}"
``` 
Also it's harder to debug if it's based on binary or Base64 so plain text based protocols are definitely preferred:

```javascript
const inner = { someProperty: "cool", otherStuff: "also cool"};
const outer = { innerProperty: JSON.stringify(inner) };
const encoder = new TextEncoder();
encoder.encode(JSON.stringify(outer));
> Uint8Array(76) [123, 34, 105, 110, 110, 101, 114, 80, 114, 111, 112, ...];
```

9. Would be nice if it had Typescript typings shipped with it (or written in Typescript to begin with) but this is definitely a low priority, optional, nice to have criteria compared to the rest of them
10. If the implementation used has different dependencies per platform (browser/NodeJS) then the build must be configured in a way to avoid bundling the dependencies for the wrong platform, e.g. let's not embed megabytes of NodeJS code in the browser build nor do it the other way around.

**Comment by suyukuoacn** [comment]
The implementation will be using stable JSON stringify implementation as it the best solution for this issue.

**Comment by petermetz** [comment]
Fixed by #246
